### ISSUE TITLE
Support LIST on auth/github/map/teams

### ISSUE THREAD

**Issue by bkrodgers** [hashicorp/vault#2239]
As far as I can tell, there's currently no way to list out what teams I've registered in the Vault Github backend.  `vault list auth/github/map/teams` returns nothing.  Can we get this added?  

I don't use the feature to register users directly under `auth/github/map/users`, but if that doesn't support LIST either, it'd be good to add there too.

**Comment by jefferai** [comment]
We can't, unfortunately. The code that GitHub is built on can't support listing in a meaningful way. Same problem we had with AppID, where we couldn't add list support until AppRole.

Someone will have to build a GitHub v2 to add listing.

**Comment by bkrodgers** [comment]
Can't it just query `vault/auth/<github-mount-id>/struct/map/teams/` in the backend?

**Comment by jefferai** [comment]
Ah. I didn't notice that GitHub isn't using a salt on the paths the way App-ID was. So yes -- this is probably doable. A PR would be welcome!

**Comment by bkrodgers** [comment]
Sure, I can take a crack at it.  Do you want to reopen this issue?

**Comment by jefferai** [comment]
That would be awesome.

**Comment by bkrodgers** [comment]
I'm having some difficulty figuring out why this isn't already working.  The team and user maps are just PathMaps.  Looking at https://github.com/hashicorp/vault/blob/master/logical/framework/path_map.go#L111, it already adds a Path for list.  

I know it's hitting this path fine, because `vault path-help auth/github/map/teams` returns:

```
Request:        map/teams
Matching Route: ^map/teams$

Read path mappings for teams


## DESCRIPTION

<no description>
```

I also know it's hitting it because `vault list auth/github/map/teams` returns `No value found at auth/github/map/teams/`, not an unsupported path error.  But if I put a debug statement above line 140 in the `pathList` function, it doesn't print out.  It seems like it's not getting into the function that path is mapped to.

I'm still playing around with it, but any thoughts you have would be appreciated.

**Comment by jefferai** [comment]
I don't have much knowledge; I believe the list support in the path map stuff was added but perhaps not actually ever used, because App-ID also salted paths which ended up making it basically useless to list anything. One obvious thing is that it should probably have a `/?` at the end to respond to list commands that do or don't contain a `/` at the end of the path.

You may want to try to ensure that the values being passed in to the path are what you are expecting. You may also want to put some debugging or a `panic` inside the list function. I wouldn't be surprised if the path that it's querying is not actually correct; if you know that you're successfully inside that function, it would tell you which of those is the issue.

**Comment by bkrodgers** [comment]
Thanks Jeff!  The `/?` tip was the first problem.  The path help still worked without it, but as the CLI list command uses `?list=true`, that ended up with a slash added somewhere.  Once that was fixed I got into the right function, but then found another issue -- it was calling the storage backend's generic `List` function, not the `PathMap.List()` function, which maps things to the proper backend path. 

I'm adding a unit test -- there was already one for `PathMap.List()`, but it wasn't testing it through the HTTP API and thus missed that it wasn't even calling that function.  Wrapping things up now, will submit a PR soon.
